### PR TITLE
[RDY] fix access violation bug in JsonIn::error()

### PR DESCRIPTION
**Pull request by ghost**
``` C++
    size_t pos = tell();
    rewind(3, 240);
    size_t startpos = tell();
    char buffer[241];
    stream->read(&buffer[0], pos - startpos);
    buffer[pos - startpos] = '\0';
```

Above was a part of code in `void JsonIn::error()`. Array `buffer` stores the context of JSON document to be shown when a JSON error occurs. The content of `buffer` was assumed to be at most 240 characters long, but in fact, in some condition the real length may exceed a little (probably because `\r` & `\n` not handled properly in `JsonIn::rewind()`? Just guess.). Today I met this bug when trying a mod by a friend.

I dynamically allocated `buffer` to avoid the problem roughly.

### CODE REVIEW

**Comment by DLaboratory** [comment]
Why the check failed?
```
Started by user the Jently bot
Building on master in workspace /var/lib/jenkins/jobs/Cataclysm-Jently/workspace
[WS-CLEANUP] Deleting project workspace...
Cloning the remote Git repository
Using shallow clone
Cloning repository git@github.com:CleverRaven/Cataclysm-DDA.git
 > /usr/bin/git init /var/lib/jenkins/jobs/Cataclysm-Jently/workspace # timeout=10
Fetching upstream changes from git@github.com:CleverRaven/Cataclysm-DDA.git
 > /usr/bin/git --version # timeout=10
 > /usr/bin/git -c core.askpass=true fetch --tags --progress git@github.com:CleverRaven/Cataclysm-DDA.git +refs/heads/*:refs/remotes/origin/* --depth=1
 > /usr/bin/git config remote.origin.url git@github.com:CleverRaven/Cataclysm-DDA.git # timeout=10
 > /usr/bin/git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
 > /usr/bin/git config remote.origin.url git@github.com:CleverRaven/Cataclysm-DDA.git # timeout=10
Fetching upstream changes from git@github.com:CleverRaven/Cataclysm-DDA.git
 > /usr/bin/git -c core.askpass=true fetch --tags --progress git@github.com:CleverRaven/Cataclysm-DDA.git +refs/pull/*:refs/remotes/origin/pr/* --depth=1
 > /usr/bin/git rev-parse refs/remotes/origin/pr/17570/merge^{commit} # timeout=10
 > /usr/bin/git rev-parse refs/remotes/origin/origin/pr/17570/merge^{commit} # timeout=10
Checking out Revision 4f77ed10aeb19ff728a0cee37b6dbf2e14ba5554 (refs/remotes/origin/pr/17570/merge)
 > /usr/bin/git config core.sparsecheckout # timeout=10
 > /usr/bin/git checkout -f 4f77ed10aeb19ff728a0cee37b6dbf2e14ba5554
First time build. Skipping changelog.
[workspace] $ /bin/zsh -ex /tmp/hudson4538703568173383394.sh
+/etc/zsh/zshenv:15> [[ -z /usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games || /usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games == /bin:/usr/bin ]]
+/tmp/hudson4538703568173383394.sh:3> i=Makefile
+/tmp/hudson4538703568173383394.sh:4> diff -q /var/lib/jenkins/cataclysm-reference-repo/Makefile Makefile
+/tmp/hudson4538703568173383394.sh:3> i=src/json.h
+/tmp/hudson4538703568173383394.sh:4> diff -q /var/lib/jenkins/cataclysm-reference-repo/src/json.h src/json.h
+/tmp/hudson4538703568173383394.sh:3> i=src/json.cpp
+/tmp/hudson4538703568173383394.sh:4> diff -q /var/lib/jenkins/cataclysm-reference-repo/src/json.cpp src/json.cpp
Files /var/lib/jenkins/cataclysm-reference-repo/src/json.cpp and src/json.cpp differ
Build step 'Execute shell' marked build as failure
[WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] Skipped based on build state FAILURE
Finished: FAILURE
```

**Comment by codemime** [comment]
> Why the check failed?

No worries, the buildbot marks all PRs that change `json.cpp` as failure.

File: src/json.cpp [hunk 1]
```diff
@@ -1474,10 +1474,11 @@ void JsonIn::error(std::string message, int offset)
     size_t pos = tell();
     rewind(3, 240);
     size_t startpos = tell();
-    char buffer[241];
-    stream->read(&buffer[0], pos - startpos);
+    char *buffer = new char[pos - startpos + 1];
```
Comment (BevapDin) [inline src/json.cpp]: Would be even better without raw memory allocation, `buffer` could be a string for example. Initialized as `std::string buffer( pos - startpos, '\0' );` - note that this does not need the terminating 0-char (and no `+ 1` to store it) as the string knows its length. It also avoids the need to `delete` the buffer.

Btw. `delete buffer;` invokes undefined behavior, the pointer was created using `new []`, and must therefor be freed using `delete []`. But using a string will neatly avoid this problem.

**Comment by mugling** [comment]
Stalled?

File: src/json.cpp (hunk 1 above)
Comment (DLaboratory) [inline src/json.cpp]: Thanks for reminding me the incorrect use of `delete` operation. 

The content of `std::string` is also allocated in heap memory, so there should be no virtual difference if I allocate the char buffer manually.

Comment (mugling) [inline src/json.cpp]: Use of `new` and `delete` in C++ is almost invariably a code smell

**Comment by mugling** [comment]
@DLaboratory doesn't look like much work to finish this. Do you want to continue or can this be closed?

**Comment by DLaboratory** [comment]
Using `new` and `delete` is fine for me, if you do not accept this fix, I'll close this.

**Comment by mugling** [comment]
Why don't you want to follow @BevapDin's suggestion?

**Comment by DLaboratory** [comment]
Well, I don't see the necessity to use `std::string` provided in STL.

**Comment by mugling** [comment]
It would prevent any later memory leak if more code was inserted that lead to `delete` being missed. Manual memory management in C++ is almost always wrong.

**Comment by DLaboratory** [comment]
But can we directly `istream::read` the stream to a string like this?
```C++
std::string buffer( pos - startpos, '\0' );
stream->read( &buffer[0], pos - startpos );
```

I remembered the content of `std::string` isn't stored in the memory contiguously in some cases.

**Comment by mugling** [comment]
In `C++11` they are guaranteed contiguous: that is writing to `&str[0]` is legal

**Comment by DLaboratory** [comment]
Done.

**Pull request merged**